### ISSUE TITLE
Unmapping <c-i> when in Vimwiki

### ISSUE THREAD

**Issue by llinfeng** [vimwiki/vimwiki#607]
I found `<c-i>` to have been remapped to jump to the next Hyperlink/`[link_text](file_link)` or `[[filename]]`. This only applies to my `*.wiki` and `*.md` files, which are all of Filetype `vimwiki`. 

I would like to unmap such jump-to-hyperlink action, and restore its original functionality. I have tried to look in the documentation, and have tried the [`:Listmaps`](https://github.com/vim-scripts/listmaps.vim) tool. Nothing relevant was found.

Please advise which parameter to tweak. Thanks a lot!

**Comment by EinfachToll** [comment]
This is a very common problem in Vim. `<C-i>` isn't explicitly mapped, but `<Tab>` is, and for Vim, these keys are the same, because that's how most terminal emulators work. I don't know if in the year 2019 there is a solution, but a quick google search suggests otherwise.
That means you can either map another key to jumping to the next link: `:nmap <Leader>wn <Plug>VimwikiNextLink` (but then, `<tab>` doesn't work anymore, too), or you bear with it.

**Comment by llinfeng** [comment]
Thanks a lot! @EinfachToll 

I bet I was revisiting an old and well-known issue :) Thank you for carefully going through the details about `<c-i>` and `<tab>`!

To note, the `nmap` setting only "wipes out" the `<tab>` mapping in Normal mode.  It is great that I can still use `<tab>` in Insert-mode to go from cell to cell in tables.

**Comment by llinfeng** [comment]
@EinfachToll May I leave the `:nmap <F21> <Plug>VimwikiNextLink`  specific to the `~/ftplugin/vimwiki.vim`? Or, such mappings can only be issued in `~/.vimrc` in order to prevent Vimwiki from mapping `<tab>` to `:<Plug>VimwikiNextLink`? 

This may be a "new" issue for the most recent version. According to my comments from last week, I should have dropped the mappings to the `~/.vim/ftplugin/vimwiki.vim`, and `<tab>` was not mapped back then.

I am almost done with the workaround :) Documenting things up and will post the walkaround here.

**Comment by llinfeng** [comment]
Here goes a rough "big picture" of my attempt to tell `<tab>` and `<c-i>` form each other in Vim Normal Mode. I will document the specific settings in the next comment. At the end of the day, I achieved the following:
* `<c-o>` and `<c-i>` are consistent regardless of where issued.
* `<tab>` key can still advance to the next URL in Vimwiki (via mapping it externally to `<F21>`, and solving the subsequent troubles.)

![image](https://user-images.githubusercontent.com/5890407/52604212-e9142000-2e37-11e9-812f-6cbf5b25b14c.png)

**Comment by llinfeng** [comment]
# Splitting Tab and <c-i>
## Description of existing behavior
* Vimwiki will assume the following mapping for `<Plug>VimwikiNextLink` should
  no other keys get assigned to it:
  ```
  nmap <tab> <Plug>VimwikiNextLink
  ```
  This get us to use the `<tab>` key to jump to the next "link" as recognized by
  the Vimwiki plugin.
* `<c-o>` and `<c-i>` are a pair of keys that helps one navigate the jump list
    * `<c-o>` goes to the pravous slot in the jump list
    * `<c-i>` goes to the next slot in the jump list, and
* Unfortunately, from the day when shell/bash/terminal/command-line has been
  first devised, `<tab>` and `<c-i>` are the same thing. Up to today, the
  path-dependency in how computer architecture evolves should get us the same
  thing --- the latest version of Vim cannot tell `<tab>` and `<c-i>` apart!
  This brings a challenge for me:
    * For years, I am used to the joy of `<tab>`-ing in Vimwiki: it is
      intuitive, and fast;
    * Also, from day-1 that I learned Vim (or read a formal book on it), I fell
      in love with the `<c-o>` and `<c-i>` pair and can not live without it :)


## Solution (through AutoHotkey and another `cmap` setting)
The goal here is delegate separate functions to `<tab>` and `<c-i>` in normal
mode. I thought it could be simple, but it took me the following 4 posts + a
random thought to eventually make it happen:

1. [Early post 1: betting too much on
   Autohotkey but was clarified on a key Vim piece](https://stackoverflow.com/questions/54524509/use-ahk-to-detect-the-vim-mode-for-gui-windows-and-configure-conditional-mappi)
   ==> inspired me to add the following mappings: devise another key for
   `<Plug>VimwikiNextLink` and thus save `<tab>` and  `<c-i>` for their jump
   list function
   ```
    " In Vim: {{{
    nmap <F13> <Plug>VimwikiNextLink
    " }}}
   ```
2. [Early post 2: bettting on Vim to update its title](https://vi.stackexchange.com/questions/18796/may-i-have-a-vim-session-report-its-mode-in-its-window-title)
   ==> can let me add mode-indicator to the "Window-Title" of the Gvim session
   in a __restrictive manner__. The trick was to tweaek the `titlestring`
   setting.
   Yet, the Vim function/parameter `mode()` there is restricted in only updating
   the title with `i` when I enter Insert mode. For all other modes, `n` will be
   in place. 
3. [Post with few view + No answer, 1](https://vi.stackexchange.com/questions/18804/may-i-let-titlestring-tell-whether-i-am-in-command-line-mode)
4. [Post with few view + No answer, 2](https://stackoverflow.com/questions/54637469/update-titlestring-in-vim-when-i-enter-leave-the-commandline)
5. Random spark:
    1. Piece 1: does `cmap` exist? YES! Then, why not to map `<F13>` back to
       `<tab>`, as I have not touched `<tab>` mapping at all.
    2. Piece 2: [this post](https://superuser.com/questions/445123/vim-sending-tab-completion-key-against-a-mapped-keystroke)
    solves the `^[` problem ==> with the bare `cmap <F13> <tab>` mapping alone, the
    `^[`-character will be inserted.
    
### Vim settings
In `~/.vimrc`, drop the following two lines:
```
" Prep step: let Vim change its title-string when in Normal mode (with limitation)
" Will update the trailing `n` to `i` when switched from normal mode to insert mode.
setglobal titlestring=%F\ %{v:servername}\ %{mode()}
" 
" Core step: <Tab> and <c-i> tweak
nmap <F21> <Plug>VimwikiNextLink
set wildcharm=<Tab>
cmap <F12> <c-i>
```

### AHK mapping
**When the window is a GVIM window in Normal mode**, remap the `<tab>` key to
`<F21>`. It is tricky to craft the precise condition. The following code are in
AutoHotKey, which used pretty advanced AHK stuff (something that I learned in
2019). Sadly, I started my first AHK script without any programming, back in
2011 or so :)

    
    ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
    ; Specify the Window-matching condition
    ;2019-01-21: updated this section, due to AHK title-matching ability in brainlessly 
    ;            chaining the title-text and the ahk_exe processname.exe 
    groupadd , GVIM             , ahk_exe gvim.exe
    groupadd , GVIM             , GVIM ahk_exe Xming.exe
    groupadd , GVIM             , Vim ahk_exe Xming.exe
    groupadd , GVIM             , | vim ahk_exe ConEmu64.exe
    groupadd , GVIM             , | vim ahk_exe Hyper.exe
    ; New attempt to add Normal-mode-identifier to the Gvim windows. 
    ; Vim settings are issued so that it adds a trailing `n` character to the Title-Text
    ; Vim setting: setglobal titlestring=%F\ %{v:servername}\ %{mode()}
    groupadd , GVIM_Normal      , md GVIM n ahk_group GVIM
    groupadd , GVIM_Normal      , md GVIM1 n ahk_group GVIM
    groupadd , GVIM_Normal      , md GVIM2 n ahk_group GVIM
    groupadd , GVIM_Normal      , md GVIM3 n ahk_group GVIM
    groupadd , GVIM_Normal      , md GVIM4 n ahk_group GVIM
    groupadd , GVIM_Normal      , md GVIM5 n ahk_group GVIM
    return
    ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
    ;
    ;;Well, an elegant mapping;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
    #IfWinActive ahk_group GVIM_Normal
        Tab:: send {F21}
    #IfWinActive
    

----

## TL;DR - Comments

### How I got into this mess

Here is how I got myself into trouble: as more and more projects require me to
write in Markdown to maintain things, and that Vimwiki can takeover the `*.md`
files with the following two settings (on top of other common Vimwiki setup
variables), I end up editing `*.md` files through Vimwiki and other work-related
files side-by-side. Then, it became confusing:
* While `<c-o>` still works when I want to go to the previous entry in the
  jump list, its buddy `<c-i>` is no longer there! Vimwiki has mapped it to jump
  to the next Vimwiki Link :)
    ```
    " Relevant settings to have Vimwiki takeover markdown files
    "
    " Wiki-specific setting: to write things with `*.md` extensions
    let wiki_1.ext = '.md'
    " Plugin "global" Setting: to have Vimwiki take over the `*.md` files
        " ==> Well, that is, when Vim opens a `*.md` file, the filetype for such
        " `*.md` file will be vimwiki (instead of markdown)
    let g:vimwiki_ext2syntax = {'.md'  : 'markdown',
                            \ '.mkd': 'markdown'}
    ```

----

### More on AHK:
More on [AutoHotKey](https://www.autohotkey.com/): It is a brilliant "system"
that is only available on Windows, and it is the very reason that I stay with
Windows machines rather than switching over to Mac. (Well, I learned this in the
hard way, by buying a MacBook Air in 2015, tried 8+ months to use it as a daily
driver, and falled back to Windows machines.)

**Comment by llinfeng** [comment]
## Excerpt of the solution
### Vim settings
In `~/.vimrc`, drop the following two lines:
```
" Prep step: let Vim change its title-string when in Normal mode (with limitation)
" Will update the trailing `n` to `i` when switched from normal mode to insert mode.
setglobal titlestring=%F\ %{v:servername}\ %{mode()}
" 
" Core step: <Tab> and <c-i> tweak
nmap <F21> <Plug>VimwikiNextLink
set wildcharm=<Tab>
cmap <F12> <c-i>
```

### AHK mapping
**When the window is a GVIM window in Normal mode**, remap the `<tab>` key to
`<F21>`. It is tricky to craft the precise condition. The following code are in
AutoHotKey, which used pretty advanced AHK stuff (something that I learned in
2019). Sadly, I started my first AHK script without any programming, back in
2011 or so :)

    
    ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
    ; Specify the Window-matching condition
    ;2019-01-21: updated this section, due to AHK title-matching ability in brainlessly 
    ;            chaining the title-text and the ahk_exe processname.exe 
    groupadd , GVIM             , ahk_exe gvim.exe
    groupadd , GVIM             , GVIM ahk_exe Xming.exe
    groupadd , GVIM             , Vim ahk_exe Xming.exe
    groupadd , GVIM             , | vim ahk_exe ConEmu64.exe
    groupadd , GVIM             , | vim ahk_exe Hyper.exe
    ; New attempt to add Normal-mode-identifier to the Gvim windows. 
    ; Vim settings are issued so that it adds a trailing `n` character to the Title-Text
    ; Vim setting: setglobal titlestring=%F\ %{v:servername}\ %{mode()}
    groupadd , GVIM_Normal      , md GVIM n ahk_group GVIM
    groupadd , GVIM_Normal      , md GVIM1 n ahk_group GVIM
    groupadd , GVIM_Normal      , md GVIM2 n ahk_group GVIM
    groupadd , GVIM_Normal      , md GVIM3 n ahk_group GVIM
    groupadd , GVIM_Normal      , md GVIM4 n ahk_group GVIM
    groupadd , GVIM_Normal      , md GVIM5 n ahk_group GVIM
    return
    ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
    ;
    ;;Well, an elegant mapping;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
    #IfWinActive ahk_group GVIM_Normal
        Tab:: send {F21}
    #IfWinActive
